### DISCUSSION TITLE
The Buffrs Registry

### DISCUSSION THREAD

**Discussion by mara-schulke** [helsing-ai/buffrs#117] [Vision]
Our vision for the Buffrs Registry is to harness its inherent positive aspects and establish it as a pivotal component in the world of Protocol Buffers. The Buffrs Registry will serve as a versatile and powerful hub for engineers and organizations alike.

This registry brings the ability to provide a community hosted _default_ registry that spans organizations, fostering collaboration and enabling seamless sharing of Protocol Buffers across projects and industries. It will facilitate a thriving open-source community, encouraging the reuse of Protocol Buffers on a global scale.

At the same time, the Buffrs Registry will be designed to be self-hostable, empowering organizations to maintain control over their APIs while promoting collaboration within their own ranks. This flexibility recognizes the diverse needs of organizations and ensures that the registry can adapt to any setup.

As a central place to publish Protocol Buffers, the Buffrs Registry will become the authoritative source for the community. Engineers can confidently publish their APIs, knowing they will be accessible to a global audience, ensuring a unified and reliable repository.

The registry will reflect distributed system properties, seamlessly handling inter-referencing of Protocol Buffers. This approach promotes modularity and efficient management of dependencies, reducing redundancy and streamlining development.

Beyond hosting Protocol Buffers, the Buffrs Registry will serve as the tool for documentation, insights, and other essential resources associated with APIs. This comprehensive API management empowers users with valuable information to make informed decisions and streamlines the development process.

To cater to varying infrastructure needs, the Buffrs Registry will support swappable storage backends, whether it's S3, local file systems, Artifactory, or others. This flexibility ensures that organizations can tailor their setup to their specific requirements.

Moreover, the Buffrs Registry will serve as the open API boundary, inviting engineers to implement their own registry proxies or implementations. This encourages extensibility and interoperability within the ecosystem, fostering innovation and integration.

In pursuing this vision, we aim to create a vibrant Protocol Buffers ecosystem that empowers engineers, streamlines API management, and fosters collaboration and innovation across organizations and industries. The Buffrs Registry will be a cornerstone of this ecosystem, facilitating the growth and adoption of Protocol Buffers as a key technology in modern software development.

**Comment by xfbs** [comment]
Progress update for the registry implementation.

Currently, there are two PRs up for review:

- [Allow building registry](https://github.com/helsing-ai/buffrs/pull/126), this fixes some issues so that the registry as it currently exists can compile.
- [Hooks up database to registry](https://github.com/helsing-ai/buffrs/pull/127), this hooks up a Postgres database using `sqlx` and allows for running migrations.

The next steps for me are to add a simple `axum` server with a `tonic` implementation for the API. The two actions that are needed at this point are:

- publish package,
- download package.
